src/databricks/labs/ucx/migration/azure_credentials.py

```diff
@@ -0,0 +1,285 @@
+    name: str
+    azure_service_principal: AzureServicePrincipal
+    created_by: str
+    read_only: bool
+    message: str
+    operation: ValidationResultOperation
+    results: list[ValidationResult]
+
+    @classmethod
+    def from_storage_credential_validation(
+        cls, storage_credential: StorageCredentialInfo, validation: ValidateStorageCredentialResponse
+    ):
+        return cls(
+            name=storage_credential.name,
+            azure_service_principal=storage_credential.azure_service_principal,
+            created_by=storage_credential.created_by,
+            read_only=storage_credential.read_only,
+            results=validation.results,
+        )
+
+
+class AzureServicePrincipalMigration:
+
+    def __init__(
+        self,
+        installation: Installation,
+        ws: WorkspaceClient,
+        azure_resource_permissions: AzureResourcePermissions,
+        azure_sp_crawler: AzureServicePrincipalCrawler,
+    ):
+        self._output_file = "azure_service_principal_migration_result.csv"
+        self._final_sp_list: list[ServicePrincipalMigrationInfo] = []
+        self._installation = installation
+        self._ws = ws
+        self._azure_resource_permissions = azure_resource_permissions
+        self._azure_sp_crawler = azure_sp_crawler
+        self._action_plan = 'service_principals_for_storage_credentials.csv'
+
+    @classmethod
+    def for_cli(cls, ws: WorkspaceClient, product='ucx'):
+        installation = Installation.current(ws, product)
+        config = installation.load(WorkspaceConfig)
+        sql_backend = StatementExecutionBackend(ws, config.warehouse_id)
+        azurerm = AzureResources(ws)
+        locations = ExternalLocations(ws, sql_backend, config.inventory_database)
+
+        azure_resource_permissions = AzureResourcePermissions(installation, ws, azurerm, locations)
+        azure_sp_crawler = AzureServicePrincipalCrawler(ws, sql_backend, config.inventory_database)
+
+        return cls(installation, ws, azure_resource_permissions, azure_sp_crawler)
+
+    def _list_storage_credentials(self) -> set[str]:
+        # list existed storage credentials that is using service principal, capture the service principal's application_id
+        storage_credential_app_ids = set()
+
+        storage_credentials = self._ws.storage_credentials.list(max_results=0)
+        for storage_credential in storage_credentials:
+            # only add service principal's application_id, ignore managed identity based storage_credential
+            if storage_credential.azure_service_principal:
+                storage_credential_app_ids.add(storage_credential.azure_service_principal.application_id)
+        logger.info(
+            f"Found {len(storage_credential_app_ids)} distinct service principals already used in UC storage credentials"
+        )
+        return storage_credential_app_ids
+
+    def _check_sp_in_storage_credentials(self, sp_list, sc_set) -> list[StoragePermissionMapping]:
```

Comment:
A little cumbersome. You can use list comprehension. Doesn't require a separate method.

Reply:
fixed